query: Increase Postgres timestamp precision

digitalcircuit opened the pull request:
## In short
* Increase [`timestamp` precision on PostgreSQL](https://www.postgresql.org/docs/current/datatype-datetime.html ) to show milliseconds
  * Makes it easier to distinguish identical messages sent next to each other
  * On pre-0.13 Quassel cores, no change
  * Additional formatting control might be useful in a separate pull request
* [SQLite is already fixed in a different pull request](https://github.com/fish-face/quasselgrep/pull/22 )
* Add `--datetime-format` to specify custom date/time format strings
  * E.g. `--datetime-format="%Y-%m-%d %H:%M:%S.%f"` shows microseconds
  * Default `%Y-%m-%d %H:%M:%S` matches current output

Criteria | Rank | Reason
---------|---------|-------------
Impact | ★☆☆ *1/3* | Adds user customization, precision to Postgres timestamps
Risk | ★☆☆ *1/3* | May break if encountering an invalid date/time
Intrusiveness | ★★☆ *2/3* | Changes to configuration and output format, might interfere with other pull requests

## Details

Since 0.13, Quassel stores timestamps with millisecond precision.  However, quasselgrep has not been showing this [due to specifying `timestamp(0)`](https://www.postgresql.org/docs/current/datatype-datetime.html ).

Including the millisecond timestamp can help with distinguishing multiple identical messages next to each other, e.g. if someone posts ASCII art or spam.

The default output remains the same.  A custom timestamp format must be specified to show milliseconds.

## Testing

### Steps

1.  Set up quasselgrep with a Quassel core v0.13 or later
2.  Ensure there's some message activity in the core database
3.  Print the messages using quasselgrep's filtering
4.  Add `--datetime-format="%Y-%m-%d %H:%M:%S.%f"` to check with microseconds

### Before, message from Quassel 0.13+
```
[2022-05-28 03:35:22] <digitalcircuit> An example message.
```

### After, message from Quassel 0.13+ *using default date/time format*
```
[2022-05-28 03:35:22] <digitalcircuit> An example message.
```

### After, message from Quassel 0.13+ with `--datetime-format`
*Using `--datetime-format="%Y-%m-%d %H:%M:%S.%f"`*
```
[2022-05-28 03:35:22.329000] <digitalcircuit> An example message.
```

digitalcircuit committed:
query: Increase Postgres timestamp precision

Increase the precision for PostgreSQL timestamp information.  Quassel
version 0.13 and higher stores timestamps in milliseconds, which makes
it easier to distinguish when multiple identical messages are sent
close together.

Messages without fractional seconds are shown as before.

Before, fractional:
[2022-05-28 03:35:22] <digitalcircuit> An example message.

After, fractional:
[2022-05-28 03:35:22.329000] <digitalcircuit> An example message.

After, message from before Quassel 0.13+:
[2014-05-28 04:19:27] <SomeUser> [Some message]

NOTE: This is using Python's default formatting.  In output.py
format() function, something like...

formatted += '[%s] ' % time.isoformat(sep=" ", timespec="milliseconds")

...would clean up the microsecond display.  However, it doesn't hurt
anything.

See https://www.postgresql.org/docs/current/datatype-datetime.html

fish-face commented:
I think I would like the default behaviour to remain outputting seconds, since (IMO) that is more common as a format in both clients and in their logs. Internally, using max available precision everywhere is fine, but I find micro (or milliseconds) too much for most human-consumed output... Would you be able to add a command-line option to do this? Either a simple switch or a format string to be interpreted by the stdlib is fine by me.

digitalcircuit commented:
@fish-face That sounds reasonable.

I'm leaning towards a [format string that's interpreted by `strftime()`](https://docs.python.org/3/library/datetime.html#datetime.datetime.strftime ) such as `--datetime-format='…'` which has a default value of the current (before this patch) date/time output format.  I imagine my use case of [taking `isoformat(sep=' ', timespec='microseconds')`](https://docs.python.org/3/library/datetime.html#datetime.datetime.isoformat ) is more niche than someone wanting to change how quasselgrep outputs date/time strings in an arbitrary way.

digitalcircuit committed:
output: Add --datetime-format to customize time

Add new option "--datetime-format" to allow specifying a custom date
and time format.  This provides a consistent default regardless of the
timestamp precision supplied by the database, and allows for
customizing the output (e.g. displaying milliseconds, AM/PM).

Example:
--datetime-format="%Y-%m-%d %H:%M:%S.%f"

See https://docs.python.org/3/library/datetime.html#datetime.datetime.strftime

digitalcircuit commented:
@fish-face I have added the `--datetime-format` option which defaults to the current quasselgrep output (seconds, `%Y-%m-%d %H:%M:%S`).

For my use case, I just need to specify `--datetime-format="%Y-%m-%d %H:%M:%S.%f"`.

This should also make it possible to [merge the SQLite pull request without impacting the timestamp display](https://github.com/fish-face/quasselgrep/pull/22) (note that SQLite timestamps has been broken since Quassel cores 0.13+).
